File: ios/Sources/AmthalPortal/ManifestClient.swift
import Foundation

/// Fetches `GET {baseURL}/embed/manifest.json` (no auth) and applies the
/// version gate of SPEC.md §5 before any route is loaded.
struct ManifestClient {
    let config: PortalConfig
    let sdkVersion: String

    /// 10 s network timeout. Gate rules:
    /// - `bridgeProtocolVersion` must equal the protocol this SDK implements.
    /// - `portalVersion` must be >= `config.minPortalVersion` (when set).
    /// - SDK version must be >= manifest `minSdkVersion` (when present).
    /// Gate failures ⇒ `versionIncompatible`; network/parse failures ⇒
    /// `manifestFailed` (except no-connectivity, surfaced as `offline` so the
    /// SDK can auto-retry when the network returns).
    func fetchAndGate() async -> Result<EmbedManifest, PortalError> {
        guard let url = config.portalURL(forPath: BridgeProtocol.manifestPath) else {
            return .failure(PortalError(
                code: .manifestFailed,
                message: "Could not construct the manifest URL from the configured base URL."
            ))
        }

        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.timeoutInterval = 10
        request.cachePolicy = .reloadIgnoringLocalCacheData

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            let nsError = error as NSError
            let offlineCodes = [
                NSURLErrorNotConnectedToInternet,
                NSURLErrorNetworkConnectionLost,
                NSURLErrorDataNotAllowed,
                NSURLErrorInternationalRoamingOff,
            ]
            if nsError.domain == NSURLErrorDomain, offlineCodes.contains(nsError.code) {
                return .failure(PortalError(
                    code: .offline,
                    message: "No internet connection while fetching the portal manifest."
                ))
            }
            return .failure(PortalError(
                code: .manifestFailed,
                message: "Manifest request failed: \(nsError.localizedDescription)"
            ))
        }

        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            let status = (response as? HTTPURLResponse)?.statusCode ?? -1
            return .failure(PortalError(
                code: .manifestFailed,
                message: "Manifest request returned HTTP \(status)."
            ))
        }

        guard let manifest = try? JSONDecoder().decode(EmbedManifest.self, from: data) else {
            return .failure(PortalError(
                code: .manifestFailed,
                message: "Manifest response could not be parsed."
            ))
        }

        if manifest.bridgeProtocolVersion != BridgeProtocol.version {
            return .failure(PortalError(
                code: .versionIncompatible,
                message: "Portal speaks bridge protocol \(manifest.bridgeProtocolVersion); "
                    + "this SDK implements protocol \(BridgeProtocol.version)."
            ))
        }

        if let minPortal = config.minPortalVersion,
           compareVersions(manifest.portalVersion, minPortal) < 0 {
            return .failure(PortalError(
                code: .versionIncompatible,
                message: "Portal version \(manifest.portalVersion) is older than the "
                    + "required minimum \(minPortal)."
            ))
        }

        if let minSdk = manifest.minSdkVersion,
           compareVersions(sdkVersion, minSdk) < 0 {
            return .failure(PortalError(
                code: .versionIncompatible,
                message: "SDK version \(sdkVersion) is older than the portal's "
                    + "required minimum \(minSdk)."
            ))
        }

        return .success(manifest)
    }
}

File: android/portal-sdk/src/main/java/com/amthalgroup/portal/internal/ManifestClient.kt
package com.amthalgroup.portal.internal

import com.amthalgroup.portal.PortalConfig
import com.amthalgroup.portal.PortalError
import com.amthalgroup.portal.PortalErrorCode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.security.cert.X509Certificate
import javax.net.ssl.HttpsURLConnection

/**
 * Fetches `GET {baseUrl}/embed/manifest.json` — against the FULL base URL
 * including any deploy path prefix (SPEC.md §2.3 "Base path awareness",
 * protocol.ts `manifestUrl()`) — and applies the version gate (SPEC.md §5).
 * No auth is attached to this request.
 */
internal object ManifestClient {

    internal data class Manifest(
        val portalVersion: String,
        val bridgeProtocolVersion: Int,
        val minSdkVersion: String?
    )

    internal sealed class Result {
        data class Success(val manifest: Manifest) : Result()
        data class Failure(val error: PortalError) : Result()
    }

    suspend fun fetchAndGate(config: PortalConfig): Result = withContext(Dispatchers.IO) {
        val raw = try {
            httpGet(config, config.normalizedBaseUrl + BridgeProtocol.MANIFEST_PATH)
        } catch (e: HttpStatusException) {
            return@withContext Result.Failure(
                PortalError(PortalErrorCode.MANIFEST_FAILED, "Manifest request failed (HTTP ${e.status})", e.status)
            )
        } catch (e: Exception) {
            return@withContext Result.Failure(
                PortalError(PortalErrorCode.MANIFEST_FAILED, e.message ?: "Manifest request failed")
            )
        }

        val manifest = try {
            val json = JSONObject(raw)
            Manifest(
                portalVersion = json.optString("portalVersion", ""),
                bridgeProtocolVersion = json.optInt("bridgeProtocolVersion", -1),
                minSdkVersion = json.optStringOrNull("minSdkVersion")
            )
        } catch (e: JSONException) {
            return@withContext Result.Failure(
                PortalError(PortalErrorCode.MANIFEST_FAILED, "Manifest is not valid JSON")
            )
        }

        gate(config, manifest)?.let { return@withContext Result.Failure(it) }
        Result.Success(manifest)
    }

    /** Returns null when compatible; a VERSION_INCOMPATIBLE error otherwise. */
    private fun gate(config: PortalConfig, manifest: Manifest): PortalError? {
        if (manifest.bridgeProtocolVersion != BridgeProtocol.PROTOCOL_VERSION) {
            return PortalError(
                PortalErrorCode.VERSION_INCOMPATIBLE,
                "Portal speaks bridge protocol ${manifest.bridgeProtocolVersion}; SDK implements ${BridgeProtocol.PROTOCOL_VERSION}"
            )
        }
        val minPortal = config.minPortalVersion
        if (minPortal != null &&
            BridgeProtocol.compareVersions(manifest.portalVersion, minPortal) < 0
        ) {
            return PortalError(
                PortalErrorCode.VERSION_INCOMPATIBLE,
                "Portal ${manifest.portalVersion} is older than required minPortalVersion $minPortal"
            )
        }
        val minSdk = manifest.minSdkVersion
        if (minSdk != null &&
            BridgeProtocol.compareVersions(BridgeProtocol.SDK_VERSION, minSdk) < 0
        ) {
            return PortalError(
                PortalErrorCode.VERSION_INCOMPATIBLE,
                "SDK ${BridgeProtocol.SDK_VERSION} is older than portal's minSdkVersion $minSdk"
            )
        }
        return null
    }

    private class HttpStatusException(val status: Int) : IOException("HTTP $status")

    private fun httpGet(config: PortalConfig, url: String): String {
        val conn = URL(url).openConnection() as HttpURLConnection
        conn.connectTimeout = 10_000
        conn.readTimeout = 10_000
        conn.requestMethod = "GET"
        conn.setRequestProperty("Accept", "application/json")
        conn.instanceFollowRedirects = true
        try {
            conn.connect()
            verifyPinning(config, conn)
            val status = conn.responseCode
            if (status < 200 || status > 299) throw HttpStatusException(status)
            return conn.inputStream.bufferedReader(Charsets.UTF_8).use { it.readText() }
        } finally {
            conn.disconnect()
        }
    }

    /** Applies the optional certificate pinning hook to an established connection. */
    internal fun verifyPinning(config: PortalConfig, conn: HttpURLConnection) {
        val pinner = config.certificatePinner ?: return
        if (conn is HttpsURLConnection) {
            val leaf = conn.serverCertificates.firstOrNull() as? X509Certificate
            if (leaf == null || !pinner(leaf)) {
                throw IOException("Certificate pinning check failed")
            }
        }
    }
}
